packages/cli/src/commands/hydrogen/preview.ts

```diff
@@ -31,47 +40,135 @@ export default class Preview extends Command {
     ...commonFlags.inspectorPort,
     ...commonFlags.debug,
     ...commonFlags.verbose,
+
+    // For building the app:
+    build: Flags.boolean({
+      description: 'Builds the app before starting the preview server.',
+    }),
+    watch: Flags.boolean({
+      description: 'Watches for changes and rebuilds the project.',
+      dependsOn: ['build'],
+    }),
+    ...overrideFlag(commonFlags.entry, {
+      entry: {dependsOn: ['build']},
+    }),
+    ...overrideFlag(commonFlags.codegen, {
+      codegen: {dependsOn: ['build']},
+    }),
+    ...overrideFlag(commonFlags.diff, {
+      diff: {dependsOn: ['build']},
+    }),
   };
 
   async run(): Promise<void> {
     const {flags} = await this.parse(Preview);
+    const originalDirectory = flags.path
+      ? resolvePath(flags.path)
+      : process.cwd();
+    let directory = originalDirectory;
+
+    if (flags.build && flags.diff) {
+      directory = await prepareDiffDirectory(originalDirectory, false);
+    }
 
-    await runPreview({
+    const {close} = await runPreview({
       ...flagsToCamelObject(flags),
+      directory,
+    });
+
+    setupResourceCleanup(async () => {
+      await close();
+      if (flags.diff) {
+        await copyDiffBuild(directory, originalDirectory);
+      }
     });
   }
 }
 
 type PreviewOptions = {
   port?: number;
-  path?: string;
+  directory?: string;
   legacyRuntime?: boolean;
   env?: string;
   envBranch?: string;
   inspectorPort?: number;
   debug: boolean;
   verbose?: boolean;
+  build?: boolean;
+  watch?: boolean;
+  entry?: string;
+  codegen?: boolean;
+  codegenConfigPath?: string;
 };
 
 export async function runPreview({
   port: appPort,
-  path: appPath,
+  directory,
   legacyRuntime = false,
   env: envHandle,
   envBranch,
   inspectorPort,
   debug,
   verbose,
+  build: shouldBuild = false,
+  watch = false,
+  codegen: useCodegen = false,
+  codegenConfigPath,
+  entry,
 }: PreviewOptions) {
-  if (!process.env.NODE_ENV) process.env.NODE_ENV = 'production';
+  if (!process.env.NODE_ENV)
+    process.env.NODE_ENV = watch ? 'development' : 'production';
```

Comment:
Using `development` with watch mode to allow our CSP to connect to Miniflare's loopback server (localhost:<random>).